rfd/00XY-session-locking.md

```diff
@@ -0,0 +1,250 @@
+
+#### `tctl` support
+
+`SessionLock` resources can be managed using `tctl [get|create|rm]`.  In this
+way, it is possible to update (e.g. delay) or remove a session lock after it
+has been created.
+
+There will be a special `tctl sessions lock` helper provided, to facilitate
+supplying time information when creating new `SessionLock`s, see Scenarios below.
+
+### Disabling new certificates
+
+No new user certificates matching a session lock target should be generated
+while the session lock is in force.  An audit event should be emitted upon such
+an attempt.
+
+A new lock check will be added to the `generateUserCert` function
+in `lib/auth/auth.go`.
+
+### Terminating existing sessions
+
+Terminating an existing session due to a (freshly created) session lock is
+similar to terminating an existing session due to certificate expiry.  The
+support for the latter is covered by `srv.Monitor` and its `Start` routine.
+
+In order to make `srv.Monitor` keep track of all the `SessionLock`s without
+periodically polling the backend, two new fields are added to `srv.MonitorConfig`:
++ `StoredSessionLocks`: a slice of `SessionLock`s known at the time of calling
+  `srv.NewMonitor`;
++ `SessionLockWatcher`: a `types.Watcher` detecting additional puts or deletes
+  of `SessionLock`s pertaining to the current session.
+
+The developed logic should work with all the proxies that already make use of
+`DisconnectExpiredCert`, i.e. SSH, k8s and DB.
+
+### Replicating to trusted clusters
+
+`SessionLock` resources are replicated from the root cluster to leaf clusters
+in a similar manner to how CAs are shared between trusted clusters.
+
+The goal should be achieved by introducing a routine similar to
+`periodicUpdateCertAuthorities`. However instead of polling the backend (with
+the default period of 10 minutes defined in `defaults.LowResPollingPeriod`)
+a `types.Watcher`-based algorithm should be preferred.
+
+To be able to distinguish session locks received from a remote (root) cluster,
+the root cluster should send the `SessionLock` resource a resource label of the
+form:
+```
+teleport.dev/replicated-from: <root-cluster-name>
+```
+
+### Scenarios
+
+#### Creating a permanent lock
+
+```
+$ tctl sessions lock --user=foo@example.com --message="Suspicious activity."
+Created a session lock with ID "dc7cee9d-fe5e-4534-a90d-db770f0234a1".
+```
+
+This locks out `foo@example.com` without automatic expiration.
+The lock can be lifted by `tctl rm lock/dc7cee9d-fe5e-4534-a90d-db770f0234a1`.
+
+The above locking command would be equivalent to
+```sh
+tctl create <<EOF
+kind: session_lock
+metadata:
+  name: dc7cee9d-fe5e-4534-a90d-db770f0234a1
+spec:
+  message: "Suspicious activity."
+  target:
+    user: foo@example.com
+version: v2
+EOF
+```
+
+The showed YAML would also correspond to the output of `tctl get lock/dc7cee9d-fe5e-4534-a90d-db770f0234a1`.
+
+#### Creating a lock with expiry
```

Comment:
Let's add a couple of targets customers have asked for:

* Teleport node was compromised. Administrator would like to put a lock on node ID or node name to prevent node from registering and heartbeating in the cluster. All sessions to that node will be terminated. The node won't be able to renew the certificate.
* A user's computer has been stolen. The computer had an MFA device registered. Place a lock on sessions initiated by certificates with device ID. (See Andrew's MFA for certificates issued with MFA metadata)